magento/magento2
# Improve cron error logging

### Description (*)
This pull request improves logging of errors within cron processes.

**Current scenario**
Currently any error messages are lost / not recorded. For the main group (and any groups not set to use a separate process), STDERR is left for the scheduler to capture. This is most easily caught with the `MAILTO=` directive on unix, however experience suggests this is rarely configured usefully. Any cron group currently configured to run in a separate process will have its STDOUT and STDERR lost forever.

**Proposed scenario**
With the changes here, both STDERR and STDOUT are caught and logged. This is very valuable in scenarios where one cannot configure the scheduler, or when a job running in a separate process is suffering from errors. A good example of this is Magento Cloud hosting where cron processes often fail without any useful diagnostic information available. These changes will mean that STDERR output is readily available, aiding in diagnostics.

### Related Pull Requests


### Fixed Issues (if relevant)
*None*

### Manual testing scenarios (*)
1. Artificially introduce a memory leak into a cron job, eg `indexer_reindex_all_invalid`
1. Run `php bin/magento cron:run` without these changes
1. Observe lack of information about the failure
1. Run `php bin/magento cron:run` with these changes
1. Observe information available in `var/log` (in this example, `var/log/magento.cron.index.log`)

### Questions or comments
I can see that [using `posix_isatty()` is discouraged](https://github.com/magento/magento-coding-standard/blob/887ecba8edaff155951094051822995cff3bca1c/Magento2/Sniffs/Functions/DiscouragedFunctionSniff.php#L107), but no alternative is provided / suggested. There are functions in packages installed via composer as development dependencies which provide the required functionality. (eg [here](https://github.com/sebastianbergmann/environment/blob/5.1.3/src/Console.php#L105) and [here](https://github.com/squizlabs/PHP_CodeSniffer/blob/3.5.8/src/Util/Common.php#L160)) What's the recommended approach in this case? Do we simply remove the output line altogether?

### Contribution checklist (*)

### Resolved issues:
1. [x] resolves magento/magento2#37453: Improve cron error logging
